src/ssr-template-helper.js

```diff
@@ -0,0 +1,134 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Capability} from './service/viewer-impl';
+
+/**
+ * Helper, that manages the proxying of template rendering to the viewer.
+ */
+export class SsrTemplateHelper {
+
+  /**
+   * @param {string} sourceComponent
+   * @param {!./service/viewer-impl.Viewer} viewer
+   * @param {!./service/templates-impl.Templates} templates
+   */
+  constructor(sourceComponent, viewer, templates) {
+
+    /** @private */
+    this.viewer_ = viewer;
+
+    /** @private */
+    this.templates_ = templates;
+
+    /** @private {!XMLSerializer} */
+    this.xmls_ = new XMLSerializer();
+
+    this.sourceComponent = sourceComponent;
+
+    /** @const @private {boolean} */
+    this.viewerCanRenderTemplate_ = this.viewer_.canRenderTemplates();
+  }
+
+  /**
+   * Whether the viewer can render templates.
+   * @return {boolean}
+   */
+  isSupported() {
+    return this.viewerCanRenderTemplate_;
+  }
+
+  /**
+   * Proxies xhr and template rendering to the viewer and renders
+   * the response.
+   * @param {!Element} element
+   *     viewer to determine the component responsible for proxying the
+   *     request.
+   * @param {?function():Promise<?>} renderTemplateSuccessCallback
+   * @param {?function():Promise<?>} renderTemplateFailureCallback
+   * return {!Promise}
+   */
+  fetchAndRenderTemplate(
+    element,
+    renderTemplateSuccessCallback,
+    renderTemplateFailureCallback) {
+    const inputsAsJson = this.getElementInputsAsJson_(element);
+    const elementAttrsAsJson = this.getElementAttributesAsJson_(element);
+    elementAttrsAsJson.inputData = inputsAsJson;
+    const mustacheTemplate = this.xmls_.serializeToString(
+        this.templates_.findTemplate(element));
+    return this.viewer_.sendMessageAwaitResponse(
+        Capability.VIEWER_RENDER_TEMPLATE,
+        {
```

Comment:
I might have missed this but can we define types for these requests and responses like what we did for xhr?  https://github.com/ampproject/amphtml/blob/master/src/service/xhr-impl.js#L52

In this way we can quickly check are refer to when implementing our viewer.

Reply:
sure thing, we can definitely define a typedef for this expected payload